### PR TITLE
Deep Cody: Allow toggle in UI & implement CodyChatMemory

### PR DESCRIPTION
**Pull request by abeatrix**
CLOSE: https://linear.app/sourcegraph/issue/CODY-4246/remove-deep-cody-from-the-model-list
CLOSE: https://linear.app/sourcegraph/issue/CODY-4247/add-toggle-for-35-sonnet

Deep Cody is now an experimental feature that users can toggle on/off with a feature flag enabled. 
This allows us to remove Deep Cody from the model drop-down list, and makes it easier to enable Deep Cody for other models in the future.

UI Change:

-  Remove Deep Cody from model list
- Stops treating Deep Cody as model
- Integrates Deep Cody toggle functionality into the chat UI
- Updates the `ContextCell` and `HumanMessageEditor` components to display and handle the Deep Cody toggle
- Adds new API methods to the `WebviewToExtensionAPI` interface to enable/disable Deep Cody
- Updates the `ChatController` to handle the Deep Cody feature flag and toggle
- Removes the automatic selection of Deep Cody as the default model
- Adds unit tests for the Deep Cody feature

Deep Cody Agent class refactor:

- Introduce `CodyChatAgent` as the base class for managing chat agents
- Implement `CodyChatMemory` to store and retrieve temporary chat context
- Refactor `CodyToolProvider` to use a registry-based approach for managing tools
- Add new tool - `MemoryTool` - a tool that allows Cody to build memory when asekd
- Improve handling of OpenContext-based tools

Add agent field to chat messages

This allows for more flexibility in enabling different chat agents.

- Added `agent` field to `ChatMessage` and `SerializedChatMessage` interfaces
- Updated `ChatController` and webview components to use the `agent` field
- Enabled Deep Cody context when the `agent` is set to `'deep-cody'`
- Updated tests and other related code to reflect the new `agent` field

## Test plan

Start branch locally and confirmed everything works as expected

image

Memory Tool

image

image

### CODE REVIEW

File: vscode/webviews/components/shadcn/ui/switch.tsx [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import * as SwitchPrimitives from '@radix-ui/react-switch'
+import * as React from 'react'
+import { cn } from '../utils'
+
+const Switch = React.forwardRef<
```
Comment (valerybugakov) [inline vscode/webviews/components/shadcn/ui/switch.tsx:5]: I noticed several accessibility issues with this component: 

1. Hovering over it triggers two popup windows overlapping each other: Screenshot 2024-11-05 at 17 24 53
2. On the light modern theme, the toggled state looks strange. I expect it to be similar to the disabled state but with a round thing on the left side. 
    Screenshot 2024-11-05 at 17 25 28
3. The toggle is not wrapped when there's not enough space for it: Screenshot 2024-11-05 at 17 29 39
4. Looks different in solarized light and dark classic themes. With the dark theme, it's hard to distinguish between enabled/disabled states.
    Screenshot 2024-11-05 at 17 32 33
    Screenshot 2024-11-05 at 17 32 27
    Screenshot 2024-11-05 at 17 31 59
    Screenshot 2024-11-05 at 17 31 52

**Review by valerybugakov — APPROVED** [review]
Tested locally, approving with some feedback inline. It works! 

Screenshot 2024-11-05 at 17 28 05

File: vscode/webviews/components/shadcn/ui/switch.tsx (hunk 1 above)
Comment (abeatrix) [reply vscode/webviews/components/shadcn/ui/switch.tsx:5]: Good catch on the double pop up. I added a title for the label that is no longer relevant, will remove it. Thanks for catching that!

File: vscode/src/chat/agentic/CodyChatMemory.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import { type ContextItem, ContextItemSource } from '@sourcegraph/cody-shared'
+import { URI } from 'vscode-uri'
+import { localStorage } from '../../services/LocalStorageProvider'
+
+export class CodyChatMemory {
+    private static Store = new Set<string>()
+
+    public static load(memory: string): void {
+        CodyChatMemory.Store.add(memory)
+        localStorage?.set(CodyChatMemory.STORAGE_KEY, Array.from(CodyChatMemory.Store))
```
[the remark this answers is not in the document]
Comment (abeatrix) [reply vscode/src/chat/agentic/CodyChatMemory.ts:10]: Ah good point. I'll limit it to 5 items to start with and we can increase them in the future if needed.

**Review by valerybugakov — COMMENTED** [review]
To align with our recent efforts to enhance release stability and enforce stricter PR reviews, I recommend breaking PRs into smaller, focused changes where possible. This approach has several benefits: if an issue arises, we can roll back only the specific change rather than the entire function, and focused changes make the review process much easier.

Since this is an experimental feature and it's already reviewed, splitting it isn't required, but let's keep this in mind for future submissions. We can aim to keep non-generated changes in PRs to a minimum necessary for the PR's primary goal.

**Comment by abeatrix** [comment]
> To align with our recent efforts to enhance release stability and enforce stricter PR reviews, I recommend breaking PRs into smaller, focused changes where possible. This approach has several benefits: if an issue arises, we can roll back only the specific change rather than the entire function, and focused changes make the review process much easier.
> 
> 
> 
> Since this is an experimental feature and it's already reviewed, splitting it isn't required, but let's keep this in mind for future submissions. We can aim to keep non-generated changes in PRs to a minimum necessary for the PR's primary goal.

Agree, and noted!
Thanks for the thorough review @valerybugakov 🙇‍♀️

File: vscode/src/chat/agentic/CodyChatMemory.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (PriNova) [reply vscode/src/chat/agentic/CodyChatMemory.ts:10]: What about storing the memory in a file (workspace or user scope)

[the remark this answers is not in the document]
Comment (abeatrix) [reply vscode/src/chat/agentic/CodyChatMemory.ts:10]: Good idea! I think that'd be super helpful as that allows users to update it manually as well 👀 added this as a follow-up item to the project. Thanks @PriNova!

File: vscode/webviews/components/shadcn/ui/switch.tsx (hunk 1 above)
Comment (taiyab) [reply vscode/webviews/components/shadcn/ui/switch.tsx:5]: Thanks for catching these @valerybugakov.

@abeatrix This makes the panel width required to be quite wide, any way we can reduce this down or allow for proper wrapping?

Comment (abeatrix) [reply vscode/webviews/components/shadcn/ui/switch.tsx:5]: I've updated to address the light theme issue. 

image

> This makes the panel width required to be quite wide, any way we can reduce this down or allow for proper wrapping?

@taiyab we can remove the icon but that probably not going to help much 🤔  any advice?

image

**Pull request merged**